Indexes made reasonably fast

alf239 opened the pull request:
Indexes are reasonably fast now; more work is necessary, but it's not urgent any more.

The summary of changes is as follows:
- NetCDFs are read in one thread now: big NetCDFs require a lot of memory, so it's very hard for GC to handle more than one
- NetCDFs are sorted from big ones to smaller ones, so that we set up memory expectations early in the processing, and by the time the memory is filled with gene data, we only read small files
- min P / max T stats are surprisingly tough to handle; hence, we allocate a big thread pull (mostly idle) to handle those: that enables us to release file handles and large data arrays sooner, helping GC to do its job. Additionally, we don't wait for heavy map juggling and just proceed reading files.
- Experiment indexing is one-threaded now: it enables further Hibernate optimisations would we need those, and it does not affect the indexing time as the slowest part is gene indexing, not experiments'

Commit:
Index on Sample.experiment

Commit:
Unused methods

Commit:
typo

Commit:
From task queue to threadlocal

Commit:
@GuardedBy clause

Commit:
Eliminated unnecessary copying

Commit:
Experimenting with parallel processing

Commit:
TreeSet -> FastSet

Commit:
One thread to read data, wide pool to process summaries

Commit:
Single-threaded Experiment index builder; less threads for Gene Index so that they would not battle for Solr locks

Commit:
Merge branch 'alf' into indexes

alf239 commented on atlas-index-api/src/main/java/uk/ac/ebi/gxa/statistics/Statistics.java:
```diff
@@ -89,6 +89,45 @@ public class Statistics implements Serializable, StatisticsBuilder {
+        scoringExperiments.addAll(other.scoringExperiments);
+    }
+
+    private <T, V, D, M extends Map<V, Map<D, ConciseSet>>> void mergeMMMC(Map<T, M> a, Map<T, M> b) {
```
the weird "MMMC" name stands for "`merge`(`Map` of `Map` of `Map` of `ConciseSet`)". The only reason we need such a name is Java's type erasure: `merge(Map<T, M> a, Map<T, M> b)`, `merge(Map<T, Map<V, ConciseSet>> a, Map<T, Map<V, ConciseSet>> b)`, and `merge(Map<T, ConciseSet> a, Map<T, ConciseSet> b)`, as we'd all prefer to call these, would all be just `merge(Map, Map)`.

Commit:
Notes on the weird naming

Commit:
Same annotation version for all modules

Commit:
Restoring (erroneously removed) imports

Commit:
Hibernate 3.5.1-Final

Commit:
Progress in GiB

Commit:
Extracted progress monitoring into a separate class in order to avoid errors like 25G/39G (100%)

alf239 commented:
Update:
- clarified tricky code, 
- fixed mistakes in progress messages
- switched to Hibernate 3.5.1-Final (mostly due to IntelliJ bug, but why not?)
- extracted some of the supplementary code so that both progress and processing logic are more clear now

Commit:
Preparing for the merge - NetCDFs do not reference to experiment ID any more

Merge remote branch 'origin/develop' into indexes

Conflicts:
	indexbuilder/src/main/java/uk/ac/ebi/gxa/index/builder/service/GeneAtlasBitIndexBuilderService.java

alf239 commented:
Merged with `develop`, ready for review. Anyone?

Commit:
Missing change, sorry

Commit:
From object indexes to real objects, part I: code migration

alf239 committed:
Getting rid of "index" in comments and variable names: they're not indexes any more

alf239 committed:
Immutable ExperimentInfo

Commit:
Naming conventions, minor readability enhancements

Commit:
Primitives for P and T

Commit:
No redundant "value" in serialized form any more

Commit:
Updated bitindex for unit tests, fixed a bug in ObjectPool (first instance of an object was always lost), eliminated large int array copy.

Commit:
Merge branch 'kiss' into indexes

alf239 commented:
Update:
- Instead of object indices, we're using real objects now: object reference takes the same 4 bytes as an `Integer` reference, so we just drop the indexes, and have code which is much more clear
- Eliminated duplicated strings from the serialized form 
- Eliminated `bioEntityIds` copying
- Made P and T primitives
- Adjusted T rank to be more similar to corresponding T values: we use these values instead of T, anyway
- Rearranged the methods to keep them closer to the corresponding data
- Made ExperimentInfo immutable and lean
- In total, won approximately 10% of disk space for the bitindex

Commit:
equals and hashCode

rpetry commented on atlas-index-api/src/main/java/uk/ac/ebi/gxa/statistics/EfoIndex.java:
```diff
@@ -1,64 +1,59 @@
+import static com.google.common.collect.Sets.newHashSet;
+
 /**
  * This class stores a mapping between efo terms and their corresponding ef-efv (i.e. Attribute)-Experiment index combinations
```
Please update javadoc to reflect changes in data stractures used by this class

alf239 replied:
done

rpetry commented on atlas-index-api/src/main/java/uk/ac/ebi/gxa/statistics/EfoIndex.java:
```diff
@@ -1,64 +1,59 @@
-    private Map<String, Map<Integer, Set<Integer>>> efoIndex = new HashMap<String, Map<Integer, Set<Integer>>>();
+    private Map<String, Map<ExperimentInfo, Set<EfvAttribute>>> efoIndex = newHashMap();
 
     // Attribute -> experiment index -> efoTerm
```
javadoc needs updating

alf239 replied:
done

rpetry commented on atlas-index-api/src/main/java/uk/ac/ebi/gxa/statistics/ThreadSafeStatisticsBuilder.java:
```diff
@@ -4,93 +4,72 @@ import com.google.common.collect.Multiset;
-    private final ExecutorService executor = new ThreadPoolExecutor(1, 1, 1, TimeUnit.SECONDS,
-            new ArrayBlockingQueue<Runnable>(100, false), new BlockOnRejectedExecutionHandler());
+    @GuardedBy("threadLocals")
+    private final List<Statistics[]> threadLocals = new ArrayList<Statistics[]>();
```
It would be really useful to see a javadoc on why we're having to use an array here

alf239 replied:
Extracted the trick in its own class, commented the latter

rpetry commented on indexbuilder/src/main/java/uk/ac/ebi/gxa/index/builder/service/ExperimentAtlasIndexBuilderService.java:
```diff
@@ -52,63 +51,33 @@ import static com.google.common.collect.Collections2.transform;
-            // if we have encountered an exception, throw the first error
-            if (firstError != null) {
-                throw new IndexBuilderException("An error occurred whilst building the Experiments index", firstError);
+            // fetch all experiments - check if we want all or only the pending ones
```
Where do we check for pending experiments?

alf239 replied:
The check has long gone — thanks for spotting.

To be more precise, it has gone in revision 74e3d382 by @pashky:

> Ditch all index builder "update pending" logic as we're not going to support it in any foreseen future + and there go all  "pending genes" APIs as there ain't such thing
> 
> git-svn-id: svn://bar.ebi.ac.uk/branches/atlas-standalone@10971 2913f559-6b04-0410-9a09-c530ee9f5186

Dropped the comment as obsolete

Commit:
BitIndex: getting rid of "index" in javadoc and variable names as a follow-up for pull request #13

Commit:
BitIndex: cleaning up comments as a follow-up for pull request #13

The check has long gone — thanks for spotting.

To be more precise, it has gone in revision 74e3d38 by @pashky:

> Ditch all index builder "update pending" logic as we're not going to support it in any foreseen future + and there go all "pending genes" APIs as there ain't such thing
>
> git-svn-id: svn://bar.ebi.ac.uk/branches/atlas-standalone@10971 2913f559-6b04-0410-9a09-c530ee9f5186

Dropped the comment as obsolete

Commit:
BitIndex: getting rid of "index" in javadoc and variable names

Commit:
BitIndex: extracting the GC-friendly threadlocal into a separate class

Commit:
BitIndex: clarifying the EFO reader code - not actually necessary, but it's  readable now

alf239 commented:
Well, indexing finishes perfectly for me. There are two other problems, though — not sure they are connected to this change, thus would like to make sure they aren't before proceeding.

alf239 commented:
First problem, `ArrayIndexOutOfBoundsException` in E-GEOD-22001, is an existing one, and can be reproduced in `master`, hence not in scope of this change (tracked as Trac ticket 2995)

Commit:
Logging improved during debug, not really necessary

Commit:
Addressing NPE: if we don't know about a particular EFO, that means the map in question is empty; no need to return `null`

alf239 commented:
Second one may or may not be introduced by the change: basically, when we draw an anatomogram (e.g. for Saa4/`ENSMUSG00000040017`), and one of the organism parts has an EFO (here `EFO_0000793`) having a child (here `EFO_0000844`) which is not present in the `EfoIndex`, we return `null` as a map from experiments to attributes, which obviously causes trouble.

The question stays open how come that we store that child in the first place.

alf239 commented:
Sorry, it was obvious: the children are taken from EFO, not from the index. Thus we cannot expect that index contains the EFO we have at hand. 

Fixed, anyway. 

@rpetry — please review.
